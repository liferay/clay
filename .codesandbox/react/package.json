{
	"version": "1.0.0",
	"name": "clay-codesandbox",
	"private": true,
	"dependencies": {
		"@clayui/alert": "latest",
		"@clayui/autocomplete": "latest",
		"@clayui/badge": "latest",
		"@clayui/breadcrumb": "latest",
		"@clayui/button": "latest",
		"@clayui/card": "latest",
		"@clayui/charts": "latest",
		"@clayui/color-picker": "latest",
		"@clayui/css": "latest",
		"@clayui/data-provider": "latest",
		"@clayui/date-picker": "latest",
		"@clayui/drop-down": "latest",
		"@clayui/form": "latest",
		"@clayui/icon": "latest",
		"@clayui/label": "latest",
		"@clayui/link": "latest",
		"@clayui/list": "latest",
		"@clayui/loading-indicator": "latest",
		"@clayui/management-toolbar": "latest",
		"@clayui/modal": "latest",
		"@clayui/multi-select": "latest",
		"@clayui/multi-step-nav": "latest",
		"@clayui/nav": "latest",
		"@clayui/navigation-bar": "latest",
		"@clayui/pagination-bar": "latest",
		"@clayui/pagination": "latest",
		"@clayui/panel": "latest",
		"@clayui/popover": "latest",
		"@clayui/progress-bar": "latest",
		"@clayui/shared": "latest",
		"@clayui/slider": "latest",
		"@clayui/sticker": "latest",
		"@clayui/table": "latest",
		"@clayui/tabs": "latest",
		"@clayui/time-picker": "latest",
		"@clayui/tooltip": "latest",
		"react": "16.8.4",
		"react-dom": "16.8.4",
		"react-scripts": "2.1.3"
	},
	"devDependencies": {
		"typescript": "3.3.3"
	},
	"scripts": {
		"start": "react-scripts start"
	}
}
