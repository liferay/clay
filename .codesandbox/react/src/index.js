import React from 'react';
import ReactDOM from 'react-dom';
import '@clayui/css/lib/css/atlas.css';

function App() {
	return (
		<div className="App">
			<h1>Hello CodeSandbox</h1>
			<h2>Add some clay components to see the magic happen!</h2>
		</div>
	);
}

const rootElement = document.getElementById('root');
ReactDOM.render(<App />, rootElement);
