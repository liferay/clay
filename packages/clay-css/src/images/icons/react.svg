<!--
* SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
* SPDX-FileCopyrightText: © 2020 Contributors to the project Clay <https://github.com/liferay/clay/graphs/contributors>
*
* SPDX-License-Identifier: BSD-3-Clause
-->
<svg viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg"><path d="M16 8.006c0-1.041-1.327-2.028-3.362-2.64.47-2.038.26-3.66-.66-4.179a1.438 1.438 0 0 0-.726-.18v-.003c-.897 0-2.071.628-3.259 1.718-1.187-1.083-2.36-1.705-3.258-1.705v.003c-.274 0-.521.058-.737.18-.916.519-1.122 2.137-.649 4.169C1.321 5.984 0 6.968 0 8.006c0 1.041 1.327 2.028 3.362 2.64-.47 2.038-.26 3.66.66 4.179.211.122.459.18.733.18.897 0 2.07-.629 3.258-1.718 1.187 1.083 2.361 1.705 3.258 1.705.274 0 .522-.058.737-.18.917-.519 1.122-2.137.65-4.169C14.678 10.031 16 9.044 16 8.006Zm-4.752-6.287c.15.003.271.032.372.086.444.25.636 1.202.486 2.426a9.73 9.73 0 0 1-.166.942 16.066 16.066 0 0 0-2.071-.349 15.698 15.698 0 0 0-1.357-1.602c1.06-.971 2.058-1.5 2.736-1.503ZM5.11 9.647c.154.256.31.506.47.75a14.93 14.93 0 0 1-1.337-.254c.12-.413.27-.84.44-1.265.134.256.274.512.427.769Zm-.877-3.778c.427-.1.877-.186 1.344-.253a17.649 17.649 0 0 0-.894 1.528 13.57 13.57 0 0 1-.45-1.275Zm.832 2.147a17.028 17.028 0 0 1 1.461-2.503 17.267 17.267 0 0 1 2.942-.003 16.59 16.59 0 0 1 1.468 2.493 17.032 17.032 0 0 1-1.461 2.503 17.251 17.251 0 0 1-2.943.003 16.59 16.59 0 0 1-1.468-2.493Zm5.824 1.621c.15-.256.29-.516.425-.772.176.43.326.859.45 1.275-.428.103-.878.19-1.344.256.16-.246.32-.5.47-.759Zm-.003-3.272a17.37 17.37 0 0 0-.47-.75c.464.068.91.151 1.338.254-.12.413-.27.84-.44 1.265a15.136 15.136 0 0 0-.428-.769Zm-1.99-1.611A20.543 20.543 0 0 0 8 4.73c-.307 0-.61.007-.907.023.294-.375.597-.718.897-1.026.304.308.607.65.907 1.026ZM4.36 1.818a.736.736 0 0 1 .375-.083v-.003c.676 0 1.677.529 2.74 1.493a15.197 15.197 0 0 0-1.347 1.599c-.737.077-1.435.196-2.074.353a9.516 9.516 0 0 1-.17-.93c-.153-1.224.036-2.176.476-2.429Zm-.82 8.136c-.323-.1-.63-.208-.914-.327C1.471 9.144.724 8.51.724 8.006c0-.503.747-1.14 1.901-1.621.28-.119.587-.225.904-.324.186.628.43 1.282.734 1.951-.3.667-.542 1.317-.724 1.942Zm1.21 4.339a.79.79 0 0 1-.37-.086c-.443-.25-.635-1.202-.485-2.426.036-.301.095-.619.166-.942.64.153 1.337.272 2.071.349.44.593.897 1.131 1.357 1.602-1.063.971-2.061 1.503-2.74 1.503Zm3.258-2.01A13.3 13.3 0 0 1 7.1 11.26c.293.012.593.022.897.022.306 0 .61-.007.906-.023-.293.375-.596.718-.896 1.026Zm3.633 1.91a.736.736 0 0 1-.375.084c-.676 0-1.677-.529-2.74-1.493.457-.471.913-1.006 1.347-1.6a15.478 15.478 0 0 0 2.074-.352c.075.324.134.635.17.933.153 1.224-.036 2.175-.476 2.429Zm1.731-4.566c-.28.119-.587.225-.903.324A15.22 15.22 0 0 0 11.734 8c.3-.667.542-1.317.724-1.942.323.1.63.208.917.327 1.154.483 1.901 1.118 1.901 1.621-.003.503-.75 1.14-1.905 1.621Z"/><path d="M7.997 9.47c.823 0 1.49-.655 1.49-1.464 0-.809-.667-1.464-1.49-1.464s-1.49.655-1.49 1.464c0 .809.667 1.464 1.49 1.464Z"/></svg>
