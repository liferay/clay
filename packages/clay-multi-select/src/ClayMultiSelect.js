import 'clay-autocomplete';
import 'clay-button';
import 'clay-label';
import {Config} from 'metal-state';
import ClayComponent from 'clay-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';

import templates from './ClayMultiSelect.soy.js';

/**
 * Metal ClayMultiSelect component.
 * @extends ClayComponent
 */
class ClayMultiSelect extends ClayComponent {
	/**
	 * Assemble the schema of the item.
	 * @param {!string} label
	 * @param {!string} value
	 * @protected
	 * @return {!Object}
	 */
	_getItemSchema(label, value) {
		return {
			label: label,
			value: value,
		};
	}

	/**
	 * Continues the propagation of the Button clicked event.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleButtonClicked(event) {
		return !this.emit({
			name: 'buttonClicked',
			originalEvent: event,
		});
	}

	/**
	 * Handle the click on the close label button.
	 * @param {!Object} data
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleCloseButtonClick(data) {
		return this._handleItemRemoved(data.target.element);
	}

	/**
	 * Continues the propagation of the data change event
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleDataChange(event) {
		return !this.emit({
			data: event.data,
			name: 'dataChange',
			originalEvent: event,
		});
	}

	/**
	 * Handle the click on the dropdown item and the propagation of the labelAdded event.
	 * @param {!Event} event
	 * @protected
	 * @return {?Boolean} If the event has been prevented or not.
	 */
	_handleDropdownItemClick(event) {
		this.filteredItems = [];
		this.refs.autocomplete.refs.input.focus();

		const value = this._performCall(this.valueLocator, event.data);
		const label = this._performCall(this.labelLocator, event.data);
		const data = this._getItemSchema(label, value);

		return this._handleItemAdded(value, data, event, 'itemSelected');
	}

	/**
	 * Handles changes in filteredItems and synchronizes with state.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleFilteredItemsChange(event) {
		if (event.newVal !== event.prevVal) {
			this.filteredItems = event.newVal;
		}
	}

	/**
	 * Continues the propagation of the itemAdded event.
	 * @param {!String} value
	 * @param {!(object|array|string)} data
	 * @param {!Event} event
	 * @param {?String} eventName
	 * @protected
	 * @return {?Boolean} If the event has been prevented or not.
	 */
	_handleItemAdded(value, data, event, eventName = 'itemAdded') {
		const label = value.toLowerCase().replace(',', '');

		if (
			label.trim() &&
			!this.selectedItems.find(
				itemSelected =>
					this._performCall(this.valueLocator, itemSelected) === label
			)
		) {
			const index = this.selectedItems.push(data);

			this.selectedItems = this.selectedItems;

			return !this.emit({
				data: {
					item: this.selectedItems[index - 1],
				},
				name: eventName,
				originalEvent: event,
			});
		} else {
			this.refs.autocomplete.refs.input.value = label;
		}
	}

	/**
	 * Handle the focus on the selected items and triggers the focused item event.
	 * @param {!Event} event
	 * @param {?Boolean} direction
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleItemFocus(event, direction) {
		if (this.selectedItems.length) {
			const {autocomplete} = this.refs;
			const items = autocomplete.element.querySelectorAll(
				'span[id="item-tag"]'
			);

			if (this._itemFocused) {
				const index = this._itemFocused.getAttribute('data-tag');
				const condition = direction
					? Number(index) - 1
					: Number(index) + 1;

				if (condition > items.length - 1) {
					this.refs.autocomplete.refs.input.focus();
					this._removeFocusedItem();
					return false;
				} else if (condition >= 0) {
					this._itemFocused = items[condition];
					this._itemFocused.focus();
				} else {
					return false;
				}
			} else {
				this._itemFocused = items[items.length - 1];
				this._itemFocused.focus();
			}

			return !this.emit({
				data: {
					item: this._itemFocused,
				},
				name: 'itemFocused',
				originalEvent: event,
			});
		}
	}

	/**
	 * Handle the focused item removal and propagating the itemRemoved event.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleItemRemoved(event) {
		const index = event.getAttribute('data-tag');
		const item = this.selectedItems[Number(index)];

		this._removeFocusedItem();
		this.selectedItems.splice(Number(index), 1);
		this.selectedItems = this.selectedItems;

		return !this.emit({
			data: {
				item,
			},
			name: 'itemRemoved',
			originalEvent: event,
		});
	}

	/**
	 * Handles input changes and propagates the queryChange event.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleOnInput(event) {
		const value = event.data.value;

		this._removeFocusedItem();

		switch (event.data.char) {
		case ',':
			return this._handleItemAdded(
				value,
				this._getItemSchema(value, value),
				event
			);
		default:
			return !this.emit({
				data: {
					value,
				},
				name: 'inputChange',
				originalEvent: event,
			});
		}
	}

	/**
	 * Handles form interactions and propagates corresponding events.
	 * @param {!Event} event
	 * @protected
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleOnKeydown(event) {
		const {value} = this.refs.autocomplete.refs.input;

		switch (event.data.key) {
		case 'Enter':
			event.preventDefault();
			if (this._itemFocused) {
				return this._handleItemRemoved(this._itemFocused);
			} else if (value && event.data.eventFromInput) {
				return this._handleItemAdded(
					value,
					this._getItemSchema(value, value),
					event
				);
			}
			break;
		case 'Backspace':
			if (!value) {
				if (!this._itemFocused) {
					return this._handleItemFocus(event);
				} else {
					return this._handleItemRemoved(this._itemFocused);
				}
			}
			break;
		case 'ArrowLeft':
			if (!value && this._itemFocused) {
				return this._handleItemFocus(event, true);
			}
			break;
		case 'ArrowRight':
			if (!value && this._itemFocused) {
				return this._handleItemFocus(event, false);
			}
			break;
		}
	}

	/**
	 * Handles data mapping.
	 * @param {!(function|string)} param
	 * @param {!Array} data
	 * @protected
	 * @return {!(string|number)}
	 */
	_performCall(param, data) {
		if (typeof param === 'function') {
			return param(data);
		}

		if (typeof data === 'string') {
			return data;
		}

		return data[param];
	}

	/**
	 * Removes the focus from the focused element.
	 * @protected
	 */
	_removeFocusedItem() {
		if (this._itemFocused) {
			this._itemFocused = null;
		}
	}

	/**
	 * @inheritDoc
	 */
	attached() {
		this._itemFocused = null;
	}

	/**
	 * @inheritDoc
	 */
	disposed() {
		this._itemFocused = null;
	}

	/**
	 * @inheritDoc
	 */
	syncSelectedItems() {
		this.refs.autocomplete.refs.input.focus();
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayMultiSelect.STATE = {
	/**
	 * Variation name to render different deltemplates.
	 * @default undefined
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?(string|undefined)}
	 */
	contentRenderer: Config.string(),

	/**
	 * Data to add to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?object}
	 */
	data: Config.object(),

	/**
	 * The array of data items that the data source contains,
	 * the URL for the data provider to request, or a function
	 * that receives the query and returns a promise with the
	 * elements.
	 * @instance
	 * @default undefined
	 * @memberof ClayDataProvider
	 * @type {!(string|object|array|function)}
	 */
	dataSource: Config.oneOfType([
		Config.array(),
		Config.func(),
		Config.object(),
		Config.string(),
	]).required(),

	/**
	 * Set the request debounce time
	 * @instance
	 * @default 200
	 * @memberof ClayDataProvider
	 * @type {?(number)}
	 */
	debounceTime: Config.number().value(200),
	
	/**
	 * Object that wires events with default listeners
	 * @default undefined
	 * @instance
	 * @memberof ClayMultiSelect
	 * @review
	 * @type {?(object|undefined)}
	 */
	defaultEventHandler: Config.object(),

	/**
	 * CSS classes to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?(string|undefined)}
	 */
	elementClasses: Config.string(),

	/**
	 * Flag to enable the filtred items with autocomplete default.
	 * @default true
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?bool}
	 */
	enableAutocomplete: Config.bool().value(true),

	/**
	 * List of filtered items for suggestion or autocomplete.
	 * @default []
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?Array}
	 */
	filteredItems: Config.array(Config.object()).value([]),

	/**
	 * Help text to guide the user in the interaction.
	 * @default undefined
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {!string}
	 */
	helpText: Config.string().required(),

	/**
	 * Id to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?(string|undefined)}
	 */
	id: Config.string(),

	/**
	 * Set some initial data while the first request is being made
	 * @instance
	 * @default undefined
	 * @memberof ClayMultiSelect
	 * @type {?(object|array)}
	 */
	initialData: Config.oneOfType([Config.object(), Config.array()]),

	/**
	 * Name for each selected item input hidden.
	 * @default selectedItems
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?(string|undefined)}
	 */
	inputName: Config.string().value('selectedItems'),

	/**
	 * Set ups the request options
	 * @instance
	 * @default undefined
	 * @memberof ClayMultiSelect
	 * @type {?(object|undefined)}
	 */
	requestOptions: Config.shapeOf({
		method: Config.string(),
		mode: Config.string(),
		cache: Config.string(),
		credentials: Config.string(),
		headers: Config.object(),
		redirect: Config.string(),
		referrer: Config.string(),
		body: Config.object(),
	}),

	/**
	 * Flag to define how often to refetch data (ms)
	 * @instance
	 * @default 0
	 * @memberof ClayMultiSelect
	 * @type {?(number|undefined)}
	 */
	requestPolling: Config.number().value(0),

	/**
	 * Define how many attempts will be made when the request fails
	 * @instance
	 * @default 5
	 * @memberof ClayMultiSelect
	 * @type {?(number|undefined)}
	 */
	requestRetries: Config.number().value(5),

	/**
	 * Set timeout of the request
	 * @instance
	 * @default 30000
	 * @memberof ClayMultiSelect
	 * @type {?(number|undefined)}
	 */
	requestTimeout: Config.number().value(30000),

	/**
	 * Label of the input element.
	 * @default undefined
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?(string|undefined)}
	 */
	label: Config.string(),

	/**
	 * Sets the name of the field to map the label of the item.
	 * @default label
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?(function|string)}
	 */
	labelLocator: Config.oneOfType([Config.func(), Config.string()]).value(
		'label'
	),

	/**
	 * List of the selected Items.
	 * @default []
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?Array<Object>}
	 */
	selectedItems: Config.array(Config.object()).value([]),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @default undefined
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {!string}
	 */
	spritemap: Config.string().required(),

	/**
	 * Sets the name of the field to map the value of the item.
	 * @default value
	 * @instance
	 * @memberof ClayMultiSelect
	 * @type {?(function|string)}
	 */
	valueLocator: Config.oneOfType([Config.func(), Config.string()]).value(
		'value'
	),
};

defineWebComponent('clay-multi-select', ClayMultiSelect);

Soy.register(ClayMultiSelect, templates);

export {ClayMultiSelect};
export default ClayMultiSelect;
