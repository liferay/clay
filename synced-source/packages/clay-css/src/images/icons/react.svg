<!--
* SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
* SPDX-FileCopyrightText: © 2020 Contributors to the project Clay <https://github.com/liferay/clay/graphs/contributors>
*
* SPDX-License-Identifier: BSD-3-Clause
-->
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
	<path class="lexicon-icon-outline" d="M512,259.8c0-33.9-42.5-66-107.6-85.9c15-66.3,8.3-119.1-21.1-136c-6.8-4-14.6-5.8-23.3-5.8V32c-28.7,0-66.3,20.4-104.3,55.9c-38-35.2-75.6-55.5-104.3-55.5v0.1c-8.8,0-16.7,1.9-23.6,5.8C98.6,55.3,92,107.9,107.2,174C42.3,194,0,226.1,0,259.8c0,33.9,42.5,66,107.6,85.9c-15,66.3-8.3,119.1,21.1,136c6.8,4,14.7,5.8,23.5,5.8c28.7,0,66.3-20.4,104.3-55.9c38,35.2,75.6,55.5,104.3,55.5c8.8,0,16.7-1.9,23.6-5.8c29.3-16.9,35.9-69.6,20.8-135.7C469.7,325.7,512,293.6,512,259.8z M359.9,55.3c4.8,0.1,8.7,1,11.9,2.8c14.2,8.1,20.3,39.1,15.5,78.9c-1.1,9.8-3,20.1-5.3,30.7c-20.5-5-42.8-8.9-66.3-11.4c-14.1-19.3-28.7-36.8-43.4-52.1C306.3,72.6,338.2,55.4,359.9,55.3z M163.5,313.2c4.9,8.3,9.9,16.5,15,24.4c-14.8-2.2-29.1-4.9-42.8-8.2c3.9-13.5,8.7-27.3,14.1-41.2C154.1,296.5,158.6,304.9,163.5,313.2z M135.5,190.3c13.7-3.2,28.1-6,43-8.2c-5.1,8-10.2,16.3-15,24.7c-4.8,8.3-9.3,16.7-13.6,25C144.2,217.8,139.4,203.8,135.5,190.3z M162.1,260.2c6.5-14,13.8-27.9,21.6-41.6c8.1-14.1,16.5-27.4,25.2-39.8c15.6-1.4,31.3-2.1,47.2-2.1c15.8,0,31.5,0.7,47,2c8.7,12.4,17.1,25.7,25.3,39.6c7.9,13.7,15.1,27.5,21.7,41.5c-6.6,14-13.8,27.9-21.6,41.6c-8.1,14.1-16.5,27.4-25.2,39.8c-15.5,1.4-31.3,2.1-47.2,2.1c-15.8,0-31.5-0.7-47-2c-8.7-12.4-17.1-25.7-25.3-39.6C175.9,288,168.6,274.1,162.1,260.2z M348.5,312.9c4.8-8.3,9.3-16.8,13.6-25.1c5.6,14,10.4,27.9,14.4,41.5c-13.7,3.3-28.1,6.2-43,8.3C338.5,329.6,343.7,321.4,348.5,312.9z M348.4,206.5c-4.8-8.3-9.9-16.5-15-24.4c14.8,2.2,29.1,4.9,42.8,8.2c-3.9,13.5-8.7,27.3-14.1,41.2C357.8,223.1,353.3,214.8,348.4,206.5z M284.7,154c-9.4-0.4-19-0.7-28.7-0.7c-9.8,0-19.5,0.2-29,0.7c9.4-12.2,19.1-23.4,28.7-33.4C265.4,130.6,275.1,141.8,284.7,154z M139.5,58.5c3.1-1.9,7.2-2.7,12-2.7v-0.1c21.6,0,53.6,17.2,87.7,48.6c-14.6,15.3-29.2,32.7-43.1,52c-23.6,2.5-45.9,6.4-66.4,11.5c-2.4-10.4-4.2-20.5-5.4-30.2C119.4,97.7,125.4,66.7,139.5,58.5z M113.2,323.2c-10.3-3.2-20.1-6.8-29.2-10.6c-36.9-15.7-60.8-36.4-60.8-52.8s23.9-37.1,60.8-52.8c9-3.9,18.8-7.3,28.9-10.5c5.9,20.4,13.8,41.7,23.5,63.5C126.8,281.7,119.1,302.9,113.2,323.2z M152,464.4c-4.7-0.1-8.7-1-11.8-2.8c-14.2-8.1-20.4-39.1-15.6-78.9c1.1-9.8,3-20.1,5.3-30.7c20.5,5,42.8,8.9,66.3,11.4c14.1,19.3,28.7,36.8,43.4,52.1C205.6,447.1,173.7,464.4,152,464.4z M256.2,399c-9.7-10-19.4-21.2-29-33.4c9.4,0.4,19,0.7,28.7,0.7c9.8,0,19.5-0.2,29-0.7C275.5,377.9,265.8,389,256.2,399z M372.5,461.2c-3.1,1.9-7.2,2.7-12,2.7c-21.6,0-53.6-17.2-87.7-48.6c14.6-15.3,29.2-32.7,43.1-52c23.6-2.5,45.9-6.4,66.4-11.5c2.4,10.5,4.3,20.6,5.4,30.3C392.6,422,386.6,453,372.5,461.2z M427.9,312.6c-9,3.9-18.8,7.3-28.9,10.5c-5.9-20.4-13.8-41.7-23.5-63.5c9.6-21.7,17.3-42.9,23.2-63.2c10.3,3.2,20.1,6.8,29.3,10.6c36.9,15.7,60.8,36.4,60.8,52.8C488.7,276.2,464.8,297,427.9,312.6z"/>
	<path class="lexicon-icon-outline" d="M255.9,307.5c26.3,0,47.7-21.3,47.7-47.7s-21.4-47.7-47.7-47.7c-26.3,0-47.7,21.3-47.7,47.7S229.6,307.5,255.9,307.5z"/>
</svg>