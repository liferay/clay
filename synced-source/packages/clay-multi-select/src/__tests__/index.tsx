/**
 * SPDX-FileCopyrightText: Â© 2019 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */

import ClayMultiSelect from '..';
import {cleanup, fireEvent, render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

global.ResizeObserver = require('resize-observer-polyfill');

const items = [
	{
		label: 'foo',
		value: '1',
	},
	{
		label: 'bar',
		value: '2',
	},
	{
		label: 'baz',
		value: '3',
	},
];

const ClayMultiSelectWithState = (props: any) => {
	const [items, setItems] = React.useState((props.items as []) || []);
	const [value, setValue] = React.useState(props.inputValue || '');

	return (
		<ClayMultiSelect
			inputValue={value}
			items={items}
			onChange={setValue}
			onItemsChange={setItems}
			sourceItems={props.sourceItems}
			spritemap="/foo/bar"
			{...props}
		/>
	);
};

describe('Interactions', () => {
	afterEach(cleanup);

	it('clicking on autocomplete item calls onItemsChange', async () => {
		const onItemsChangeFn = jest.fn();

		const {findByRole, getByRole} = render(
			<ClayMultiSelectWithState
				items={[items[0]]}
				onItemsChange={onItemsChangeFn}
				sourceItems={items}
				spritemap="/foo/bar"
			/>
		);

		const combobox = getByRole('combobox');

		userEvent.click(combobox);

		expect(onItemsChangeFn).not.toHaveBeenCalled();

		const barOption = await findByRole('option', {name: 'bar'});

		userEvent.click(barOption);

		expect(onItemsChangeFn).toHaveBeenCalled();
	});

	it('clicking on already selected item should not calls onItemsChange if duplicated items are not allowed', async () => {
		const onItemsChangeFn = jest.fn();

		const {findByRole, getByRole} = render(
			<ClayMultiSelectWithState
				allowDuplicateValues={false}
				items={[items[0]]}
				onItemsChange={onItemsChangeFn}
				sourceItems={items}
				spritemap="/foo/bar"
			/>
		);

		const combobox = getByRole('combobox');

		userEvent.click(combobox);

		expect(onItemsChangeFn).not.toHaveBeenCalled();

		const fooOption = await findByRole('option', {name: 'foo'});

		userEvent.click(fooOption);

		expect(onItemsChangeFn).not.toHaveBeenCalled();
	});

	it('clicking the Clear All button removes items and text in the input', () => {
		const onItemsChangeFn = jest.fn();
		const onChangeFn = jest.fn();

		const {getByTitle} = render(
			<ClayMultiSelectWithState
				inputValue="foo"
				items={[items[0]]}
				onChange={onChangeFn}
				onItemsChange={onItemsChangeFn}
				sourceItems={items}
				spritemap="/foo/bar"
			/>
		);

		const clearButton = getByTitle('Clear All');

		fireEvent.click(clearButton, {});

		expect(onItemsChangeFn).toHaveBeenCalledWith([]);
		expect(onChangeFn).toHaveBeenCalledWith('');
	});

	it('clicking the Clear All button should move the focus to the input', () => {
		const {container, getByTitle} = render(
			<ClayMultiSelectWithState
				inputValue="foo"
				items={[items[0]]}
				onChange={jest.fn()}
				onItemsChange={jest.fn()}
				sourceItems={items}
				spritemap="/foo/bar"
			/>
		);

		const clearButton = getByTitle('Clear All');
		const input = container.querySelector('.form-control-inset');

		expect(document.activeElement).not.toBe(input);

		fireEvent.click(clearButton, {});

		expect(document.activeElement).toBe(input);
	});

	it('Use custom function for filtering autocomplete results', () => {
		function myFilter(items: any, value: string) {
			return items.filter(
				(item: any) =>
					item.label.match(value) || item.value.match(value)
			);
		}

		const WithCustomFilter = () => {
			const [value, setValue] = React.useState('');
			const [selectedItems, setSelectedItems] = React.useState<any>([]);

			return (
				<ClayMultiSelect
					inputValue={value}
					items={selectedItems}
					onChange={setValue}
					onItemsChange={setSelectedItems}
					sourceItems={myFilter(items, value)}
					spritemap="/foo/bar"
				/>
			);
		};

		const {container} = render(<WithCustomFilter />);

		const input = container.querySelector('input');

		fireEvent.change(input as HTMLInputElement, {
			target: {value: 'bar'},
		});

		expect(
			document.querySelector('.autocomplete-dropdown-menu')
		).toMatchSnapshot();

		fireEvent.change(input as HTMLInputElement, {
			target: {value: '1'},
		});

		expect(
			document.querySelector('.autocomplete-dropdown-menu')
		).toMatchSnapshot();
	});

	xit('allows for async sourceItems', () => {
		jest.useFakeTimers();
		const callbackFn = jest.fn();

		const AsyncMultiSelect = () => {
			const [value, setValue] = React.useState('');
			const [selectedItems, setSelectedItems] = React.useState<any>([]);
			const [isLoading, setIsLoading] = React.useState(false);

			function asyncData(query: string) {
				setIsLoading(true);

				const timer = setTimeout(() => {
					callbackFn(query);

					setIsLoading(false);
				}, 2000);

				timer.unref();
			}

			return (
				<ClayMultiSelect
					inputValue={value}
					isLoading={isLoading}
					items={selectedItems}
					onChange={(value: string) => {
						setValue(value);

						asyncData(value);
					}}
					onItemsChange={setSelectedItems}
					sourceItems={items}
					spritemap="/foo/bar"
				/>
			);
		};

		const {container} = render(<AsyncMultiSelect />);

		const input = container.querySelector('input');

		fireEvent.change(input as HTMLInputElement, {
			target: {value: 'b'},
		});

		expect(callbackFn).not.toHaveBeenCalled();

		jest.runAllTimers();

		expect(callbackFn).toHaveBeenCalledWith('b');

		jest.useRealTimers();
	});
});
