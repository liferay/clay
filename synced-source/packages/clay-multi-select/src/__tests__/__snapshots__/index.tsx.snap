// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interactions Use custom function for filtering autocomplete results 1`] = `null`;

exports[`Interactions Use custom function for filtering autocomplete results 2`] = `null`;
